[GStreamer] ASSERTs when handling flush events

philn opened the pull request:
#### a158e9cdddf4f17a2fea032e1cee52c25519bf32
<pre>
[GStreamer] ASSERTs when handling flush events
<a href="https://bugs.webkit.org/show_bug.cgi?id=288936">https://bugs.webkit.org/show_bug.cgi?id=288936</a>
&lt;<a href="https://rdar.apple.com/problem/145955450">rdar://problem/145955450</a>&gt;

Reviewed by Alicia Boya Garcia.

We need to operate from the main thread when handling flush events because it is a requirement for
usage of AbortableTaskQueue start/finishAborting.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoSinkCommon.cpp:
(WebKitVideoSinkProbe::handleFlushEvent):
(WebKitVideoSinkProbe::doProbe):

Canonical link: <a href="https://commits.webkit.org/292777@main">https://commits.webkit.org/292777@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d20625dfe406eab42b3566e67deaf64e9c8e976

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97068 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102149 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47593 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99113 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25141 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/73948 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31158 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100071 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12817 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/87793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54285 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12568 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5650 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46923 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104171 "Built successfully") | 
| | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24142 "Hash 9d20625d for PR 43110 does not build (failure)") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/24401 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83901 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82392 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/20720 "The change is no longer eligible for processing.") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/4629 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17647 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24107 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29258 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23930 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27242 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25503 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc reviewed (commented):
@philn For the sake of analysis, what thread did the flush come from and what triggered it?

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/GStreamerVideoSinkCommon.cpp:
```diff
@@ -44,31 +44,46 @@ class WebKitVideoSinkProbe {
-                } else
-                    GST_DEBUG_OBJECT(pad, "Received FLUSH_STOP without a FLUSH_START, ignoring.");
-            }
+            callOnMainThreadAndWait([&] {
```
Let's add a comment so that the next person that comes here knows why this is needed. For example:

```
// Usually flushes propagate in the main thread as a synchronous consequence of a seek.
// However, this doesn't have to be the case:
//
// As a notable example, when matroskademux receives a seek before it has parsed the
// entire file header, it stores the event and returns without flushing anything.
// Later, the streaming thread finishes parsing the file header and handles the stored
// seek event from that same thread. This sends a flush from the streaming thread.
```

ntrrgc reviewed (approved):
LGTM.

philn committed:
[GStreamer] ASSERTs when handling flush events
https://bugs.webkit.org/show_bug.cgi?id=288936
<rdar://problem/145955450>

Reviewed by Alicia Boya Garcia.

We need to operate from the main thread when handling flush events because it is a requirement for
usage of AbortableTaskQueue start/finishAborting.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoSinkCommon.cpp:
(WebKitVideoSinkProbe::handleFlushEvent):
(WebKitVideoSinkProbe::doProbe):

Canonical link: https://commits.webkit.org/292777@main
